src/jshint.js

```diff
@@ -1732,32 +1732,20 @@ var JSHINT = (function() {
 
   /*
    * read all directives
-   * recognizes a simple form of asi, but always
-   * warns, if it is used
    */
   function directives() {
     var i, p, pn;
 
     while (state.tokens.next.id === "(string)") {
-      p = peek(0);
-      if (p.id === "(endline)") {
-        i = 1;
-        do {
-          pn = peek(i++);
-        } while (pn.id === "(endline)");
-        if (pn.id === ";") {
-          p = pn;
-        } else if (pn.value === "[" || pn.value === ".") {
-          // string -> [ | . is a valid production
-          break;
-        } else if (!state.option.asi || pn.value === "(") {
-          // string -> ( is not a valid production
-          warning("W033", state.tokens.next);
-        }
-      } else if (p.id === "." || p.id === "[") {
+      pn = peek();
+      i = 0;
+      do {
+        p = pn;
+        pn = peek(++i);
+      } while (p.id === "(endline)");
+      if (p.lbp > 0 &&
+          (startLine(pn) !== p.line || !checkPunctuators(p, ["++", "--"]))) {
```

Comment:
Are `++` and `--` the only tokens which may be not related to the previous one if they are on a different line?